pkg/koordlet/resourceexecutor/resctrl_updater.go

```diff
@@ -0,0 +1,184 @@
+	klog.V(6).Infof("generate new resctrl mba schemata resource, file %s, key %s, value %s",
+		schemataFile, mbSchemataKey, schemata.MBString())
+
+	return &ResctrlSchemataResourceUpdater{
+		DefaultResourceUpdater: DefaultResourceUpdater{
+			key:        mbSchemataKey,
+			file:       schemataFile,
+			value:      schemata.MBString(),
+			updateFunc: UpdateResctrlSchemataFunc,
+		},
+		schemataRaw: schemata,
+	}
+}
+
+func CalculateResctrlL3TasksResource(group string, taskIds []int32) (ResourceUpdater, error) {
+	// join ids into updater value and make the id updates one by one
+	tasksPath := sysutil.GetResctrlTasksFilePath(group)
+
+	// use ordered slice
+	sort.Slice(taskIds, func(i, j int) bool {
+		return taskIds[i] < taskIds[j]
+	})
+	var builder strings.Builder
+	for _, id := range taskIds {
+		builder.WriteString(strconv.FormatInt(int64(id), 10))
+		builder.WriteByte('\n')
+	}
+
+	return NewCommonDefaultUpdaterWithUpdateFunc(tasksPath, tasksPath, builder.String(), UpdateResctrlTasksFunc)
+}
+
+func UpdateResctrlSchemataFunc(u ResourceUpdater) error {
+	r, ok := u.(*ResctrlSchemataResourceUpdater)
+	if !ok {
+		return fmt.Errorf("not a ResctrlSchemataResourceUpdater")
+	}
+
+	schemataFile := r.Path()
+
+	oldR, err := sysutil.ReadResctrlSchemataRaw(schemataFile, r.schemataRaw.L3Number())
+	if err != nil {
+		return fmt.Errorf("failed to read current resctrl schemata, path %s, err: %v", schemataFile, err)
+	}
+
+	isEqual, msg := oldR.Equal(r.schemataRaw)
+	if isEqual { // schemata unchanged, no need to update
+		klog.V(6).Infof("skip update resctrl schemata, old l3 %s, mba %s, new %s, l3Num %v",
+			oldR.L3String(), oldR.MBString(), r.Value(), r.schemataRaw.L3Number())
+		return nil
+	}
+	klog.V(5).Infof("need to update resctrl schemata, old l3 %s, mba %s, new %s, l3Num %v, msg: %s",
+		oldR.L3String(), oldR.MBString(), r.Value(), r.schemataRaw.L3Number(), msg)
+
+	// NOTE: currently, only l3 and mba schemata are to update, so do not read or compare before the write
+	// eg.
+	// $ cat /sys/fs/resctrl/schemata/BE/schemata
+	// L3:0=7ff;1=7ff
+	// MB:0=100;1=100
+	// $ echo "L3:0=3f;1=3f" > /sys/fs/resctrl/BE/schemata
+	// $ cat /sys/fs/resctrl/BE/schemata
+	// L3:0=03f;1=03f
+	// MB:0=100;1=100
+	_ = audit.V(5).Reason(ReasonUpdateResctrl).Message("update %v to %v", u.Key(), u.Value()).Do()
+	return sysutil.CommonFileWrite(u.Path(), u.Value())
+}
+
+func UpdateResctrlTasksFunc(resource ResourceUpdater) error {
+	// NOTE: resctrl/{...}/tasks file is required to appending write a task id once a time, and any duplicate would be
+	//       dropped automatically without an exception
+	// eg.
+	// $ echo 123 > /sys/fs/resctrl/BE/tasks
+	// $ echo 124 > /sys/fs/resctrl/BE/tasks
+	// $ echo 123 > /sys/fs/resctrl/BE/tasks
+	// $ echo 122 > /sys/fs/resctrl/BE/tasks
+	// $ tail -n 3 /sys/fs/resctrl/BE/tasks
+	// 122
+	// 123
+	// 124
+	info := resource.(*DefaultResourceUpdater)
```

Comment:
also here, use method of interface directly

Reply:
fixed